Code/Framework/AzCore/AzCore/DOM/DomValue.cpp

```diff
@@ -0,0 +1,1241 @@
+    }
+
+    Object::Iterator Value::FindMutableMember(KeyType name)
+    {
+        Object::ContainerType& object = GetObjectInternal();
+        return AZStd::find_if(
+            object.begin(), object.end(),
+            [&name](const Object::EntryType& entry)
+            {
+                return entry.first == name;
+            });
+    }
+
+    Object::Iterator Value::FindMutableMember(AZStd::string_view name)
+    {
+        return FindMutableMember(AZ::Name(name));
+    }
+
+    Value& Value::MemberReserve(size_t newCapacity)
+    {
+        GetObjectInternal().reserve(newCapacity);
+        return *this;
+    }
+
+    bool Value::HasMember(KeyType name) const
+    {
+        return FindMember(name) != GetObjectInternal().end();
+    }
+
+    bool Value::HasMember(AZStd::string_view name) const
+    {
+        return HasMember(AZ::Name(name));
+    }
+
+    Value& Value::AddMember(KeyType name, const Value& value)
+    {
+        Object::ContainerType& object = GetObjectInternal();
+        object.reserve((object.size() / Object::ReserveIncrement + 1) * Object::ReserveIncrement);
+        if (auto memberIt = FindMutableMember(name); memberIt != object.end())
+        {
+            memberIt->second = value;
+        }
+        else
+        {
+            object.emplace_back(name, value);
+        }
+        return *this;
+    }
+
+    Value& Value::AddMember(AZStd::string_view name, const Value& value)
+    {
+        return AddMember(AZ::Name(name), value);
+    }
+
+    Value& Value::AddMember(AZ::Name name, Value&& value)
+    {
+        Object::ContainerType& object = GetObjectInternal();
+        if (auto memberIt = FindMutableMember(name); memberIt != object.end())
+        {
+            memberIt->second = value;
+        }
+        else
+        {
+            object.emplace_back(name, value);
+        }
+        return *this;
+    }
+
+    Value& Value::AddMember(AZStd::string_view name, Value&& value)
+    {
+        return AddMember(AZ::Name(name), value);
+    }
+
+    void Value::RemoveAllMembers()
+    {
+        GetObjectInternal().clear();
+    }
+
+    void Value::RemoveMember(KeyType name)
+    {
+        Object::ContainerType& object = GetObjectInternal();
+        object.erase(AZStd::remove_if(
+            object.begin(), object.end(),
+            [&name](const Object::EntryType& entry)
+            {
+                return entry.first == name;
+            }));
+    }
+
+    void Value::RemoveMember(AZStd::string_view name)
+    {
+        RemoveMember(AZ::Name(name));
+    }
+
+    Object::Iterator Value::RemoveMember(Object::Iterator pos)
+    {
+        Object::ContainerType& object = GetObjectInternal();
+        Object::Iterator nextIndex = object.end();
+        auto lastEntry = object.end() - 1;
+        if (pos != lastEntry)
+        {
+            AZStd::swap(*pos, *lastEntry);
+            nextIndex = pos;
+        }
+        object.resize(object.size() - 1);
+        return nextIndex;
+    }
```

Comment:
Is there protection against this being called with `pos` equal to the end iterator?

Also idiomatically, what should be used is `object.pop_back()` instead of resize.

Finally the return value of this function doesn't make sense and is invalid.
If `pos`  is not the last element in the Object container, then it just swaps with the the it returns that position iterator, which contains the value that was swapped, not the "nextIndex".
Futhermore if `pos` is the last entry in the Object container, then nextIndex is returning an invalid iterator, since the `object.resize()` would have updated the end iterator.

```suggestion
        Object::ContainerType& object = GetObjectInternal();
        if (!object.empty())
        {
            AZStd::swap(*pos, object.back());
            object.pop_back();
        }
       return object.end()
    }
```

Reply:
Used your suggestion verbatim, cheers!

As for why I was returning the old index... *cough* it involved me misreading the docs on the standard for `std::remove`